cmd/snap-confine/seccomp-support.c

```diff
@@ -17,869 +17,140 @@
-	//   - seccomp syscall
-	//   - prctl with PR_SET_SECCOMP
-	//   - ptrace (trace) in AppArmor
-	//   - capability sys_admin in AppArmor
-	// Note that with NO_NEW_PRIVS disabled, CAP_SYS_ADMIN is required to
-	// change the seccomp sandbox.
-
-	if (getresuid(&real_uid, &effective_uid, &saved_uid) != 0)
-		die("could not find user IDs");
-
-	// If running privileged or capable of raising, disable nnp
-	if (real_uid == 0 || effective_uid == 0 || saved_uid == 0)
-		if (seccomp_attr_set(ctx, SCMP_FLTATR_CTL_NNP, 0) != 0)
-			die("Cannot disable nnp");
-
-	// Note that secure_gettenv will always return NULL when suid, so
-	// SNAPPY_LAUNCHER_SECCOMP_PROFILE_DIR can't be (ab)used in that case.
-	if (secure_getenv("SNAPPY_LAUNCHER_SECCOMP_PROFILE_DIR") != NULL)
-		filter_profile_dir =
-		    secure_getenv("SNAPPY_LAUNCHER_SECCOMP_PROFILE_DIR");
+	debug("loading bpf program for security tag %s", filter_profile);
 
 	char profile_path[512];	// arbitrary path name limit
-	sc_must_snprintf(profile_path, sizeof(profile_path), "%s/%s",
+	sc_must_snprintf(profile_path, sizeof(profile_path), "%s/%s.bpf",
 			 filter_profile_dir, filter_profile);
 
-	f = fopen(profile_path, "r");
-	if (f == NULL) {
-		fprintf(stderr, "Can not open %s (%s)\n", profile_path,
-			strerror(errno));
-		die("aborting");
-	}
-	// Note, preprocess_filter() die()s on error
-	preprocess_filter(f, &pre);
-
-	if (pre.unrestricted) {
-		seccomp_release(ctx);
-		ctx = NULL;
-		goto out;
-	}
-	// FIXME: right now complain mode is the equivalent to unrestricted.
-	// We'll want to change this once we seccomp logging is in order.
-	if (pre.complain) {
-		seccomp_release(ctx);
-		ctx = NULL;
-		goto out;
-	}
-
-	char buf[SC_MAX_LINE_LENGTH];
-	while (fgets(buf, sizeof(buf), f) != NULL) {
-		lineno++;
-
-		// skip policy-irrelevant lines
-		if (validate_and_trim_line(buf, sizeof(buf), lineno) == 0)
-			continue;
-
-		char *buf_copy = strdup(buf);
-		if (buf_copy == NULL)
-			die("Out of memory");
-
-		int pr_rc = parse_line(buf_copy, &sargs);
-		free(buf_copy);
-		if (pr_rc != PARSE_OK) {
-			// as this is a syscall whitelist an invalid syscall
-			// is ok and the error can be ignored
-			if (pr_rc == PARSE_INVALID_SYSCALL)
-				continue;
-			die("could not parse line");
-		}
-
-		rc = seccomp_rule_add_exact_array(ctx, SCMP_ACT_ALLOW,
-						  sargs.syscall_nr,
-						  sargs.length, sargs.arg_cmp);
-		if (rc != 0) {
-			rc = seccomp_rule_add_array(ctx, SCMP_ACT_ALLOW,
-						    sargs.syscall_nr,
-						    sargs.length,
-						    sargs.arg_cmp);
-			if (rc != 0) {
-				fprintf(stderr,
-					"seccomp_rule_add_array failed with %i for '%s'\n",
-					rc, buf);
-				errno = 0;
-				die("aborting");
-			}
-		}
-	}
-
- out:
-	if (f != NULL) {
-		if (fclose(f) != 0)
-			die("could not close seccomp file");
-	}
-	sc_map_destroy();
-	return ctx;
-}
-
-void sc_load_seccomp_context(scmp_filter_ctx ctx)
-{
-	int rc;
+	// validate profile-path and all parents is all root owned
+	// (TOCTTOU sadly).
```

Comment:
Note, that validate_bpfpath_is_safe() assumes an absolute path, which is guaranteed in sc_apply_seccomp_bpf() but a quick check for:
```
if (path == NULL || strlen(path) == 0 || path[0] != '/')
    die(...)
```

would be good.